File: Threema/VoIPSoundViewController.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2013-2020 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "VoIPSoundViewController.h"
#import "VoIPSounds.h"
#import "UserSettings.h"
#import "BundleUtil.h"

#import <AVFoundation/AVFoundation.h>

@interface VoIPSoundViewController ()

@end

@implementation VoIPSoundViewController {
    NSArray *voIPSounds;
    NSIndexPath *selectedIndexPath;
    AVAudioPlayer *_audioPlayer;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    voIPSounds = [VoIPSounds getVoIPSounds];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationDidEnterBackground:) name:UIApplicationDidEnterBackgroundNotification object:nil];
    
    [self.tableView reloadData];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [_audioPlayer stop];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidEnterBackgroundNotification object:nil];
}

- (void)playVoIPSound:(NSString*)voIPSoundName {
    if ([voIPSoundName isEqualToString:@"default"]) {
        return;
    }
    
    NSString *soundPath = [BundleUtil pathForResource:voIPSoundName ofType:@"caf"];
    NSURL *soundUrl = [NSURL fileURLWithPath:soundPath];
    
    if (_audioPlayer) {
        if ([_audioPlayer.url isEqual:soundUrl]) {
            if (_audioPlayer.isPlaying) {
                [_audioPlayer stop];
            } else {
                _audioPlayer.currentTime = 0;
                [_audioPlayer play];
            }
            return;
        }
        [_audioPlayer stop];
        _audioPlayer = nil;
    }
    
    _audioPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:soundUrl error:nil];
    _audioPlayer.numberOfLoops = 2;
    [_audioPlayer play];
}

- (NSString*)currentVoIPSound {
    return [UserSettings sharedUserSettings].voIPSound;
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return voIPSounds.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"SoundCell"];
    
    NSString *soundName = voIPSounds[indexPath.row];
    NSString *soundNameLoc = [NSString stringWithFormat:@"sound_%@", soundName];
    cell.textLabel.text = NSLocalizedString(soundNameLoc, nil);
    
    if ([self.currentVoIPSound isEqualToString:soundName]) {
        selectedIndexPath = indexPath;
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    } else {
        cell.accessoryType = UITableViewCellAccessoryNone;
    }
    
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [UserSettings sharedUserSettings].voIPSound = voIPSounds[indexPath.row];
//    [[CallManager sharedInstance] createNewProvider];
    
    [self playVoIPSound:voIPSounds[indexPath.row]];
    
    if (selectedIndexPath != nil)
        [self.tableView cellForRowAtIndexPath:selectedIndexPath].accessoryType = UITableViewCellAccessoryNone;
    
    [self.tableView cellForRowAtIndexPath:indexPath].accessoryType = UITableViewCellAccessoryCheckmark;
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    selectedIndexPath = indexPath;
}


#pragma mark - Notifications

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    if (_audioPlayer) {
        [_audioPlayer stop];
    };
}

@end

File: Threema/Settings/VoIPSoundViewController.swift
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2021-2023 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import CocoaLumberjackSwift
import Foundation

class VoIPSoundViewController: ThemedTableViewController {
    
    private let voIPSounds = [
        "default",
        "threema_best",
        "threema_incom",
        "threema_xylo",
        "threema_goody",
        "threema_alphorn",
    ]
    private var audioPlayer: AVAudioPlayer?
    private var selectedIndexPath: IndexPath?
    private var businessInjector = BusinessInjector()
    
    // MARK: - Lifecycle
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        title = BundleUtil.localizedString(forKey: "settings_threema_calls_call_sound")
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(applicationDidEnterBackground(application:)),
            name: UIApplication.didEnterBackgroundNotification,
            object: nil
        )
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        audioPlayer?.stop()
        
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didEnterBackgroundNotification,
            object: nil
        )
    }
    
    // MARK: - UITableViewDataSource
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        voIPSounds.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = self.tableView.dequeueReusableCell(withIdentifier: "SoundCell", for: indexPath)
        let soundName = voIPSounds[indexPath.row]
        let soundNameLoccalizationKey = "sound_\(soundName)"
        
        cell.textLabel?.text = BundleUtil.localizedString(forKey: soundNameLoccalizationKey)
        
        if businessInjector.userSettings.voIPSound == soundName {
            selectedIndexPath = indexPath
            cell.accessoryType = .checkmark
        }
        else {
            cell.accessoryType = .none
        }
        return cell
    }
    
    // MARK: - UITableViewDelegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        businessInjector.userSettings.voIPSound = voIPSounds[indexPath.row]
        
        playVoIPSound(voIPSoundName: voIPSounds[indexPath.row])
        
        if let selectedIndexPath = selectedIndexPath {
            tableView.cellForRow(at: selectedIndexPath)?.accessoryType = .none
        }
        tableView.cellForRow(at: indexPath)?.accessoryType = .checkmark
        tableView.deselectRow(at: indexPath, animated: true)
        selectedIndexPath = indexPath
    }
    
    // MARK: - Notifications
    
    @objc private func applicationDidEnterBackground(application: UIApplication) {
        audioPlayer?.stop()
    }
    
    // MARK: - Helper methods
    
    private func playVoIPSound(voIPSoundName: String) {
        
        guard voIPSoundName != "default" else {
            audioPlayer?.stop()
            return
        }
        
        guard let soundPath = BundleUtil.path(forResource: voIPSoundName, ofType: "caf") else {
            DDLogError("Unable to load sound file path for `\(voIPSoundName)`")
            audioPlayer?.stop()
            return
        }
        let soundURL = URL(fileURLWithPath: soundPath)
        
        if let audioplayer = audioPlayer {
            if audioplayer.url == soundURL {
                if audioplayer.isPlaying {
                    audioplayer.stop()
                }
                else {
                    audioplayer.currentTime = 0
                    audioplayer.play()
                }
                return
            }
            audioplayer.stop()
        }
        
        do {
            audioPlayer = try AVAudioPlayer(contentsOf: soundURL)
            audioPlayer?.numberOfLoops = 2
            audioPlayer?.play()
        }
        catch {
            DDLogError("Unable to load audio player to play VoIP sound: \(error)")
        }
    }
}
